test/e2e/apps/statefulset.go

```diff
@@ -510,6 +512,155 @@ var _ = SIGDescribe("StatefulSet", func() {
+			ss, err = updateStatefulSetWithRetries(ctx, c, ns, ss.Name, func(update *appsv1.StatefulSet) {
+				update.Spec.Template.Spec.Containers[0].Image = newImage
+			})
+			framework.ExpectNoError(err)
+
+			ginkgo.By("Creating a new revision")
+			ss = waitForStatus(ctx, c, ss)
+			currentRevision, updateRevision = ss.Status.CurrentRevision, ss.Status.UpdateRevision
+			gomega.Expect(currentRevision).ToNot(gomega.Equal(updateRevision), "Current revision should not equal update revision during rolling update")
+
+			ginkgo.By("Await for all replicas running, all are updated but pod0")
+			e2estatefulset.WaitForState(ctx, c, ss, func(set2 *appsv1.StatefulSet, pods2 *v1.PodList) (bool, error) {
+				ss = set2
+				pods = pods2
+				if ss.Status.UpdatedReplicas == *ss.Spec.Replicas-1 && ss.Status.Replicas == *ss.Spec.Replicas && ss.Status.ReadyReplicas == *ss.Spec.Replicas {
+					// rolling updated is not completed, because replica 0 isn't ready
+					return true, nil
+				}
+				return false, nil
+			})
+
+			ginkgo.By("Verify pod images before pod0 deletion and recreation")
+			for i := range pods.Items {
+				if i < int(*ss.Spec.UpdateStrategy.RollingUpdate.Partition) {
+					gomega.Expect(pods.Items[i].Spec.Containers[0].Image).To(gomega.Equal(oldImage), fmt.Sprintf("Pod %s/%s has image %s not equal to oldimage image %s",
+						pods.Items[i].Namespace,
+						pods.Items[i].Name,
+						pods.Items[i].Spec.Containers[0].Image,
+						oldImage))
+					gomega.Expect(pods.Items[i].Labels[appsv1.StatefulSetRevisionLabel]).To(gomega.Equal(currentRevision), fmt.Sprintf("Pod %s/%s has revision %s not equal to current revision %s",
+						pods.Items[i].Namespace,
+						pods.Items[i].Name,
+						pods.Items[i].Labels[appsv1.StatefulSetRevisionLabel],
+						currentRevision))
+				} else {
+					gomega.Expect(pods.Items[i].Spec.Containers[0].Image).To(gomega.Equal(newImage), fmt.Sprintf("Pod %s/%s has image %s not equal to new image  %s",
+						pods.Items[i].Namespace,
+						pods.Items[i].Name,
+						pods.Items[i].Spec.Containers[0].Image,
+						newImage))
+					gomega.Expect(pods.Items[i].Labels[appsv1.StatefulSetRevisionLabel]).To(gomega.Equal(updateRevision), fmt.Sprintf("Pod %s/%s has revision %s not equal to new revision %s",
+						pods.Items[i].Namespace,
+						pods.Items[i].Name,
+						pods.Items[i].Labels[appsv1.StatefulSetRevisionLabel],
+						updateRevision))
+				}
+			}
+
+			ginkgo.By("Marking pod0 as Failed so that StatefulSet controller recreates it")
+			err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+				pod0, err := c.CoreV1().Pods(ss.Namespace).Get(ctx, pod0.Name, metav1.GetOptions{})
+				framework.ExpectNoError(err)
+				pod0.Status.Phase = v1.PodFailed
+				_, err = c.CoreV1().Pods(ss.Namespace).UpdateStatus(ctx, pod0, metav1.UpdateOptions{})
+				return err
+			})
+			framework.ExpectNoError(err)
+
+			ginkgo.By("Delete the old pod0 so that it is not counted in status")
```

Comment:
```suggestion
			ginkgo.By("Deleting the pod so that kubelet terminates it and StatefulSet controller recreates it")
```
Note that, it is not deleted immediately after calling `deleteStatefulPodAtIndex` due to the finalizer.

Reply:
changed it as suggested